### PR TITLE
Feature: Add SkyHanni User Luck to stats breakdown

### PR DESCRIPTION
**Pull request by martimavocado**
## What
Adds the SkyHanni User Luck stat to the stats breakdown in `/sbmenu`, `/equipment` and "Misc Stats" in it's submenu.
Follows the same visibility rules as other Hypixel stats. Example: Only visible when on a non default value, respects "Toggle show all stats".
Adds a custom stats breakdown menu for SkyHanni User Luck, has support for future new sources, but would probably need a bit of extra work for a new sub-sub menu for a future skills breakdown.

very much needs a new stat description, maybe a new item to represent the general stat. 

[video showcase](https://youtu.be/Qhf3-17INKg) 
screenshots are a bit outdated and don't include the new menu
image
image
image

## Changelog New Features
+ Added SkyHanni User Luck to the stats breakdown. - martimavocado
    * This can be viewed in /sbmenu, /equipment, and the Misc Stats submenu.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+            val itemStack = input[slot]
+            if (itemStack?.name == " ") {
+                return slot
+            }
+        }
+        return -1
+    }
+
+    @SubscribeEvent
+    fun onHoverItem(event: LorenzToolTipEvent) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt:153]: function too long. please extract

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    private val luckTooltipString = "§5§o §a✴ SkyHanni User Luck §f"
+    private var inCustomBreakdown = false
+
+    private val validItemSlots = (10..53).filter { it !in listOf(17, 18, 26, 27, 35, 36) && it !in 44..53 }
+    private val invalidItemSlots = (0..53).filter { it !in validItemSlots }
+
+    private var skillOverflowLuck = mutableMapOf<SkillType, Int>()
+
+    @SubscribeEvent
+    fun replaceItem(event: ReplaceItemEvent) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt:73]: function too long. please extract

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        skillsItem = Utils.createItemStack(
+            skillsID.getItemStack().item,
+            skillsName,
+            *createItemLore("skills"),
+        )
+    }
+
+    private fun createItemLore(type: String, luckInput: Float = 0.0f): Array<String> {
+        calcSkillLuck()
+        when (type) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]: you could move "return" statements before "when"

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                    "§8Action",
+                    "",
+                    "§7Value: §a+$luckString✴",
+                    "",
+                    "§8Gain more by going to Limbo,",
+                    "§8and obtaining a higher Personal Best§8.",
+                )
+            }
+
+            "skills" -> {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt:300]: extract those 25 lines into a separate function please

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    private lateinit var limboItem: ItemStack
+    private var limboID = "ENDER_PEARL".asInternalName()
+    private val limboName = "§a✴ Limbo Personal Best"
+
+    private lateinit var skillsItem: ItemStack
+    private var skillsID = "DIAMOND_SWORD".asInternalName()
+    private val skillsName = "§a✴ Category: Skills"
+
+    private var showAllStats = true
+    private val showAllStatsPattern = RepoPattern.pattern(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]: please add regex test

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+                val lastIndex = event.toolTip.indexOfLast { it == "§5§o" }
+                if (lastIndex == -1) return
+
+                val skillLuck = skillOverflowLuck.values.sum()
+                val totalLuck = skillLuck + limboLuck
+                val luckString = tryTruncateFloat(totalLuck)
+                event.toolTip.add(lastIndex, "$luckTooltipString$luckString")
+            }
+
+            else -> return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]: this can go

**Review by hannibal002 — CHANGES_REQUESTED** [review]
missing `addSeparators()`
<details>
<summary>Images</summary>

grafik
grafik


</details>

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+                createItems()
+            }
+            checkItemSlot(event)
+        }
+    }
+
+    private fun checkItemSlot(event: ReplaceItemEvent) {
+        when (event.slot) {
+            48 -> return
+            49 -> return
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]: A bit more compact.
```suggestion
            48, 49 -> {}
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+            10 -> {
+                event.replace(skillsItem)
+                return
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]: These seem unnecessary.
```suggestion
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+            }
+
+            11 -> {
+                event.replace(limboItem)
+                return
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+            }
+
+            in validItemSlots -> {
+                event.replace(null)
+                return
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+        }
+        val inventoryName = event.inventoryItems[4]?.name ?: ""
+        if (inventoryName != "§dMisc Stats") return
+        inMiscStats = true
+        replaceSlot = findValidSlot(event.inventoryItems)
+        val isAdvancedLore = event.inventoryItems[50]?.getLore() ?: listOf("")
+        isAdvancedLore.forEach {
+            val matcher = showAllStatsPattern.value.matcher(it)
+            if (matcher.find()) {
+                showAllStats = when (matcher.group("toggle")) {
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
        val showAllStatsLore = event.inventoryItems[50]?.getLore() ?: listOf("")
        for (line of showAllStatsLore) {
            showAllStatsPattern.matchMatcher(it) {
                showAllStats = when (group("toggle")) {
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+import io.github.moulberry.notenoughupdates.util.Utils
+import net.minecraft.client.player.inventory.ContainerLocalMenu
+import net.minecraft.item.ItemStack
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.time.Duration.Companion.seconds
+
+@SkyHanniModule
+object UserLuckBreakdown {
+    private var inMiscStats = false
+    private var replaceSlot = -1
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
    private var replaceSlot: Int? = null
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+                        "§7Flat: §a+$luckString✴",
+                        "",
+                        "§8You have none of this stat!",
+                        "§eClick to view!",
+                    )
+                } else {
+                    arrayOf(
+                        "§7SkyHanni User Luck increases",
+                        "§7your overall fortune",
+                        "§7fortune around Hypixel Skyblock.",
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
                        "§7fortune around Hypixel SkyBlock.",
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+        return
+    }
+
+    @SubscribeEvent
+    fun closeInventory(event: InventoryCloseEvent) {
+        inMiscStats = false
+        inCustomBreakdown = false
+    }
+
+    private fun findValidSlot(input: Map<Int, ItemStack>): Int {
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
    private fun findValidSlot(input: Map<Int, ItemStack>): Int? {
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+        for (slot in input.keys) {
+            if (slot !in validItemSlots && slot < 44) continue
+            val itemStack = input[slot]
+            if (itemStack?.name == " ") {
+                return slot
+            }
+        }
+        return -1
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/features/misc/UserLuckBreakdown.kt]:
```suggestion
        return null
```

**Review by lunaynx — CHANGES_REQUESTED** [review]

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**